### PR TITLE
YOLACT

### PR DESCRIPTION
**Pull request by chongzhou96**

### CODE REVIEW

File: mmdet/models/dense_heads/yolact_head.py [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+    3. YOLACT box head uses OHEM instead of Focal loss.
+    4. YOLACT box head predicts a set of mask coefficients for each box.
+
+    Args:
+        num_head_convs (int): Number of the conv layers shared by
+            box and cls branches.
+        num_protos (int): Number of the mask coefficients.
+        use_OHEM (bool): If true, `loss_single_OHEM` will be used for
+            cls loss calculation. If false, `loss_single` will be used.
+    """
```
Comment (yhcao6) [inline mmdet/models/dense_heads/yolact_head.py:32]: Now we follow a strict criteria of the docstring. It is necessary to annotate every parameter and every function, but you can give a simple comment in one statement like [here](https://github.com/open-mmlab/mmdetection/blob/c342781873fd8574fe9e5be279a0bf26e0c5c5a5/mmdet/core/mask/structures.py#L186) if the function is too simple.

File: mmdet/models/dense_heads/yolact_head.py [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+            if isinstance(m, nn.Conv2d):
+                xavier_init(m, distribution='uniform')
+
+    def forward(self, x, coeff_pred, bboxes, img_meta, sampling_results=None):
+        prototypes = self.protonet(x)
+        prototypes = prototypes.permute(0, 2, 3, 1).contiguous()
+
+        num_imgs = x.size(0)
+        # Training state
+        if sampling_results is not None:
```
Comment (yhcao6) [inline mmdet/models/dense_heads/yolact_head.py:497]: If I am not wrong, here we use `sampling_results is None` to decide if the model is training. But we can also use `self.training` to decide if is training or not. It will be more straightforward.

File: mmdet/models/dense_heads/yolact_head.py [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
+
+            gt_bboxes_for_reweight = cur_gt_bboxes[pos_assigned_gt_inds]
+
+            mask_h, mask_w = cur_mask_pred.shape[-2:]
+            cur_gt_masks = F.interpolate(
+                cur_gt_masks.unsqueeze(0), (mask_h, mask_w),
+                mode='bilinear',
+                align_corners=False).squeeze(0)
+            cur_gt_masks = cur_gt_masks.gt(0.5).float()
+            mask_targets = cur_gt_masks[pos_assigned_gt_inds]
```
Comment (yhcao6) [inline mmdet/models/dense_heads/yolact_head.py:574]: Similar to `YolactSegmHead` I think it will make the code more clear if we move the target computation to a new function `get_target` though it is simple.

File: configs/yolact/yolact_r50.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
+# model settings
```
[the remark this answers is not in the document]
Comment (chongzhou96) [reply configs/yolact/yolact_r50.py:1]: As YOLACT uses a different image size with what is used in coco_instance_dataset, it cannot inherit that base. However, to remove duplication, I've inherited the default_runtime.py.

**Comment by yhcao6** [comment]
Could you add unit tests for yolact head? You can refer [test_head.py](https://github.com/open-mmlab/mmdetection/blob/master/tests/test_models/test_heads.py).

File: mmdet/models/detectors/yolact.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+                      img_metas,
+                      gt_bboxes,
+                      gt_labels,
+                      gt_bboxes_ignore=None,
+                      gt_masks=None):
+        """
+        Args:
+            img (Tensor): of shape (N, C, H, W) encoding input images.
+                Typically these should be mean centered and std scaled.
+
```
Comment (yhcao6) [inline mmdet/models/detectors/yolact.py:65]: Remove the spaceline

**Comment by yhcao6** [comment]
Dear @chongzhou96, I think we have almost done but some problems about the unit tests. The coverage rate of `yolact_head.py` is 70.93%, but `yolact.py` is only 33.33%, you can check [here](https://codecov.io/gh/open-mmlab/mmdetection/pull/3456/tree?path=mmdet%2Fmodels%2Fdetectors). So I think it better to add one more unit tests in [`test_forward`](https://github.com/open-mmlab/mmdetection/blob/master/tests/test_models/test_forward.py) to test `forward` method of `Yolact` detector.

File: mmdet/core/post_processing/bbox_nms.py [hunk 6]
```diff
@@ -54,3 +56,71 @@ def multiclass_nms(multi_bboxes,
...
+             multi_scores,
+             multi_coeffs,
+             score_thr,
+             nms_cfg,
+             max_num=-1):
+    """Fast NMS in `YOLACT <https://arxiv.org/abs/1904.02689>`_.
+
+    Args:
+        multi_bboxes (Tensor): shape (n, #class*4) or (n, 4)
+        multi_scores (Tensor): shape (n, #class), where the last column
```
Comment (hellock) [inline mmdet/core/post_processing/bbox_nms.py:71]: Should the shape be (n, #class+1)?

File: mmdet/models/detectors/yolact.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    def init_segm_mask_weights(self):
+        """Initialize weights of the YOLACT semg head and YOLACT mask head."""
+        self.segm_head.init_weights()
+        self.mask_head.init_weights()
+
+    def process_gt_single(self, gt_bboxes, gt_labels, gt_masks, min_gt_box_wh,
+                          device):
+        """Cuda the gt_masks and discard boxes that are smaller than we'd
+        like."""
+        gt_masks = torch.from_numpy(gt_masks.masks).to(device)
```
Comment (hellock) [inline mmdet/models/detectors/yolact.py:38]: `gt_masks.masks` may not be a bitmap (but polygons). Use `gt_masks.to_tensor` instead.

File: mmdet/models/detectors/yolact.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        w = gt_bboxes[:, 2] - gt_bboxes[:, 0]
+        h = gt_bboxes[:, 3] - gt_bboxes[:, 1]
+        keep = (w > min_gt_box_wh[0]) * (h > min_gt_box_wh[1])
```
Comment (hellock) [inline mmdet/models/detectors/yolact.py:41]: Use `&` instead of `*`.

File: mmdet/models/detectors/yolact.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        gt_bboxes = gt_bboxes[keep]
+        gt_labels = gt_labels[keep]
+        gt_masks = gt_masks[keep]
+        return gt_bboxes, gt_labels, gt_masks
```
Comment (hellock) [inline mmdet/models/detectors/yolact.py:45]: Shouldn't these samples be filtered in the dataset?

File: mmdet/models/dense_heads/yolact_head.py [hunk 10]
```diff
@@ -0,0 +1,931 @@
...
+from ..losses import smooth_l1_loss
+from .anchor_head import AnchorHead
+
+
+@HEADS.register_module()
+class YolactHead(AnchorHead):
+    """YOLACT box head used in https://arxiv.org/abs/1904.02689.
+
+    Note that YOLACT head is a light version of RetinaNet head.
+    Four differences are described as follows:
```
Comment (hellock) [inline mmdet/models/dense_heads/yolact_head.py:20]: Add an empty line before the list to make it correctly parsed by sphinx.

File: mmdet/models/dense_heads/yolact_head.py [hunk 11]
```diff
@@ -0,0 +1,931 @@
...
+    4. YOLACT box head predicts a set of mask coefficients for each box.
+
+    Args:
+        num_classes (int): Number of categories excluding the background
+            category.
+        in_channels (int): Number of channels in the input feature map.
+        num_head_convs (int): Number of the conv layers shared by
+            box and cls branches.
+        num_protos (int): Number of the mask coefficients.
+        use_OHEM (bool): If true, `loss_single_OHEM` will be used for
```
Comment (hellock) [inline mmdet/models/dense_heads/yolact_head.py:33]: use_ohem

File: mmdet/models/dense_heads/yolact_head.py [hunk 12]
```diff
@@ -0,0 +1,931 @@
...
+            cls loss calculation. If false, `loss_single` will be used.
+        conv_cfg (dict): Dictionary to construct and config conv layer.
+        norm_cfg (dict): Dictionary to construct and config norm layer.
+        loss_cls (dict): Config of classification loss.
+        loss_bbox (dict): Config of localization loss.
+        anchor_generator (dict): Config dict for anchor generator.
```
Comment (hellock) [inline mmdet/models/dense_heads/yolact_head.py:39]: Follow the argument order of `AnchorHead`.

File: mmdet/models/dense_heads/yolact_head.py [hunk 13]
```diff
@@ -0,0 +1,931 @@
...
+                    padding=1,
+                    conv_cfg=self.conv_cfg,
+                    norm_cfg=self.norm_cfg))
+        self.yolact_cls = nn.Conv2d(
+            self.feat_channels,
+            self.num_anchors * self.cls_out_channels,
+            3,
+            padding=1)
+        self.yolact_reg = nn.Conv2d(
+            self.feat_channels, self.num_anchors * 4, 3, padding=1)
```
Comment (hellock) [inline mmdet/models/dense_heads/yolact_head.py:102]: Use unified variable names `conv_cls` and `conv_reg`.

File: mmdet/models/dense_heads/yolact_head.py [hunk 14]
```diff
@@ -0,0 +1,931 @@
...
+    def loss(self,
+             cls_scores,
+             bbox_preds,
+             gt_bboxes,
+             gt_labels,
+             img_metas,
+             gt_bboxes_ignore=None):
+        """A combination of the func:`AnchorHead.loss` and func:`SSDHead.loss`.
+
+        When self.use_OHEM == True, it functions like `SSDHead.loss`,
```
Comment (hellock) [inline mmdet/models/dense_heads/yolact_head.py:149]: The rst syntax uses double \`: When \`\`self.use_OHEM == True\`\`, it functions like \`\`SSDHead.loss\`\`

File: mmdet/models/dense_heads/yolact_head.py [hunk 15]
```diff
@@ -0,0 +1,931 @@
...
+                num_total_samples=num_total_samples)
+
+        return dict(
+            loss_cls=losses_cls, loss_bbox=losses_bbox), sampling_results
+
+    def loss_single_OHEM(self, cls_score, bbox_pred, anchors, labels,
+                         label_weights, bbox_targets, bbox_weights,
+                         num_total_samples):
+        """"See func:`SSDHead.loss`."""
+        loss_cls_all = F.cross_entropy(
```
Comment (hellock) [inline mmdet/models/dense_heads/yolact_head.py:264]: Losses are hard-coded instead of using `self.loss_cls` and `self.loss_reg`.

File: mmdet/models/detectors/yolact.py (hunk 9 above)
Comment (chongzhou96) [reply mmdet/models/detectors/yolact.py:45]: Thanks for your comments. I've finished the code modification except for this one. Correct me if I'm wrong but I don't think mmdet provides the annotation-level sample filter. Are you suggesting to let me implement such a function? If so, which file shall I add to?  Thanks.

File: configs/yolact/README.md [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+#### For a real-time demo, check out our ICCV video:
+[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/0pMfmo8qfpQ/0.jpg)](https://www.youtube.com/watch?v=0pMfmo8qfpQ)
+
+# Evaluation
+Here are our YOLACT models along with their FPS on a Titan Xp and mAP on COCO's `val`:
+
+| Image Size | Backbone      | *FPS  | mAP  | Weights                                                                                                             |
+|:----------:|:-------------:|:----:|:----:|----------------------------------------------------------------------------------------------------------------------|
+| 550        | Resnet50-FPN  | 42.5 | 29.0 | [yolact_r50_epoch_55.pth](https://drive.google.com/file/d/1S30dWxxF1jmz7Tbh1cWbHZUiMpNOK1ts/view?usp=sharing) |
+| 550        | Resnet101-FPN | 33.5 | 30.4 | [yolact_r101_epoch_55.pth](https://drive.google.com/file/d/1yv752K659KbR9arOVD1f6Rst8i00OZnX/view?usp=sharing)|
```
Comment (yhcao6) [inline configs/yolact/README.md:24]: @chongzhou96 I have trained yolact r50 myself and the mAP is exactly the same with README.md. But single GPU training takes me around 5 days, so may I ask if you have tried SyncBN or 8-GPUs training but increasing the learning rate?
And I benchmarked the inference speed but I found it hard to get the same FPS with README.md. 
I use [`benchmark.py`](https://github.com/open-mmlab/mmdetection/blob/master/tools/benchmark.py) and it should exclude the unnecessary time of data loading and post-processing.
For `Yolact resnet 50`, on v100 I can get only 17.5 FPS. May I ask have you tried to benchmark the influence speed with this Pr?

File: mmdet/core/post_processing/bbox_nms.py [hunk 17]
```diff
@@ -54,3 +56,77 @@ def multiclass_nms(multi_bboxes,
...
+
+
+def fast_nms(multi_bboxes,
+             multi_scores,
+             multi_coeffs,
+             score_thr,
+             iou_thr,
+             top_k,
+             max_num=-1):
+    """Fast NMS in `YOLACT <https://arxiv.org/abs/1904.02689>`_. Fast NMS
```
Comment (hellock) [inline mmdet/core/post_processing/bbox_nms.py:68]: """Fast NMS in \`YOLACT <https://arxiv.org/abs/1904.02689>\`_.

Fast NMS ...
"""

File: mmdet/models/dense_heads/__init__.py [hunk 18]
```diff
@@ -18,11 +18,13 @@
 from .retina_sepbn_head import RetinaSepBNHead
 from .rpn_head import RPNHead
 from .ssd_head import SSDHead
+from .yolact_head import YolactHead, YolactProtonet, YolactSegmHead
```
Comment (hellock) [inline mmdet/models/dense_heads/__init__.py:21]: Use upper case: YOLACTHead, YOLACTProtonet, YOLACTSegmHead

File: mmdet/models/detectors/yolact.py [hunk 19]
```diff
@@ -0,0 +1,149 @@
+import torch
+
+from mmdet.core import bbox2result, multi_apply
+from ..builder import DETECTORS, build_head
+from .single_stage import SingleStageDetector
+
+
+@DETECTORS.register_module()
+class Yolact(SingleStageDetector):
```
Comment (hellock) [inline mmdet/models/detectors/yolact.py:9]: YOLACT

File: mmdet/models/dense_heads/yolact_head.py [hunk 20]
```diff
@@ -0,0 +1,931 @@
...
+            _x2 = _x2.long()
+        x1 = torch.min(_x1, _x2)
+        x2 = torch.max(_x1, _x2)
+        x1 = torch.clamp(x1 - padding, min=0)
+        x2 = torch.clamp(x2 + padding, max=img_size)
+
+        return x1, x2
+
+
+class InterpolateModule(nn.Module):
```
Comment (hellock) [inline mmdet/models/dense_heads/yolact_head.py:917]: Seems no need to define it.

File: mmdet/models/dense_heads/yolact_head.py [hunk 21]
```diff
@@ -0,0 +1,931 @@
...
+        num_protos (int): Number of prototypes.
+        num_classes (int): Number of categories excluding the background
+            category.
+        loss_mask_weight (float): Reweight the mask loss by this factor.
+        max_masks_to_train (int): Maximum number of masks to train for
+            each image.
+    """
+
+    def __init__(self,
+                 protonet_cfg,
```
Comment (hellock) [inline mmdet/models/dense_heads/yolact_head.py:587]: Flatten it instead of using a dict.

File: mmdet/models/dense_heads/yolact_head.py [hunk 22]
```diff
@@ -0,0 +1,931 @@
...
+        Warning: this does things in-place behind the scenes so
+        copy if necessary.
+
+        Args:
+            _x1 (Tensor): shape (N, ).
+            _x2 (Tensor): shape (N, ).
+            img_size (int): Size of the input image.
+            padding (int): x1 >= padding, x2 <= image_size-padding.
+            cast (bool): If cast is false, the result won't be cast to longs.
+        Returns:
```
Comment (hellock) [inline mmdet/models/dense_heads/yolact_head.py:899]: Add a empty line before `Returns`

File: mmdet/models/dense_heads/yolact_head.py [hunk 23]
```diff
@@ -0,0 +1,931 @@
...
+    Args:
+        in_channels (int): Number of channels in the input feature map.
+        num_classes (int): Number of categories excluding the background
+            category.
+        loss_segm (dict): Config of semantic segmentation loss.
+    """
+
+    def __init__(self,
+                 in_channels=256,
+                 num_classes=80,
```
Comment (hellock) [inline mmdet/models/dense_heads/yolact_head.py:470]: Make `num_classes` a positional argument.

File: mmdet/models/dense_heads/yolact_head.py [hunk 24]
```diff
@@ -0,0 +1,931 @@
...
+        loss_mask_weight (float): Reweight the mask loss by this factor.
+        max_masks_to_train (int): Maximum number of masks to train for
+            each image.
+    """
+
+    def __init__(self,
+                 protonet_cfg,
+                 in_channels=256,
+                 num_protos=32,
+                 num_classes=80,
```
Comment (hellock) [inline mmdet/models/dense_heads/yolact_head.py:590]: Make `num_classes` a positional argument.

File: mmdet/models/dense_heads/yolact_head.py [hunk 25]
```diff
@@ -0,0 +1,931 @@
...
+
+        # Use sum to concat together all the component layer lists
+        net = sum([
+            make_layer(x, y) for x, y in zip(protonet_cfg['num_channels'],
+                                             protonet_cfg['kernel_size'])
+        ], [])
+        if not include_last_relu:
+            net = net[:-1]
+
+        return nn.Sequential(*(net))
```
Comment (hellock) [inline mmdet/models/dense_heads/yolact_head.py:645]: *net.

File: mmdet/models/dense_heads/yolact_head.py [hunk 26]
```diff
@@ -0,0 +1,931 @@
...
+        masks_left = rows >= x1.view(1, 1, -1)
+        masks_right = rows < x2.view(1, 1, -1)
+        masks_up = cols >= y1.view(1, 1, -1)
+        masks_down = cols < y2.view(1, 1, -1)
+
+        crop_mask = masks_left * masks_right * masks_up * masks_down
+
+        return masks * crop_mask.float()
+
+    def sanitize_coordinates(self, _x1, _x2, img_size, padding=0, cast=True):
```
Comment (hellock) [inline mmdet/models/dense_heads/yolact_head.py:885]: Do not add `_` in arguments.

**Comment by OpenMMLab-Assistant-007** [comment]
Hi!
@chongzhou96 
First of all, we want to express our gratitude for your significant PR in the OpenMMLab project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
